src/index.js

```diff
@@ -883,34 +929,49 @@ export default function (babel) {
       },
 
       ArrayComprehension(path) {
-        // disallow Yield and Return
-        path.get("loop.body").traverse({
-          AwaitExpression(awaitPath) {
-            throw awaitPath.buildCodeFrameError(
-              "`await` is not allowed within Comprehensions; " +
-              "instead, await the Comprehension (eg; `y <- [for x of xs: x]`)."
-            );
-          },
-          YieldExpression(yieldPath) {
-            throw yieldPath.buildCodeFrameError("`yield` is not allowed within Comprehensions.");
-          },
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
+
+        const id = path.scope.generateUidIdentifier("arr");
+        visitTailExpressions(path.get("loop"), true, function(tailPath) {
+          let arrPush = t.callExpression(
+            t.memberExpression(id, t.identifier("push")),
+            [tailPath.node.expression]
+          );
+          tailPath.replaceWith(arrPush);
         });
 
-        const arrId = path.scope.generateUidIdentifier("arr");
+        path.replaceWith(wrapComprehensionInIFE(id, t.arrayExpression(), path.node.loop));
+      },
 
-        transformTerminalExpressionsIntoArrPush(path.get("loop"), arrId);
+      ObjectComprehension(path) {
+        validateComprehensionLoopBody(path.get("loop.body"));
 
-        const fn = t.arrowFunctionExpression([], t.blockStatement([
-          t.variableDeclaration("const", [t.variableDeclarator(arrId, t.arrayExpression())]),
-          path.node.loop,
-          t.returnStatement(arrId),
-        ]));
+        const id = path.scope.generateUidIdentifier("obj");
+        visitTailExpressions(path.get("loop"), true, function(tailPath) {
+          const seqExpr = tailPath.node.expression;
 
-        const iife = t.callExpression(fn, []);
-        path.replaceWith(iife);
+          // Only SeqExprs of length 2 are valid.
+          if (
+            (seqExpr.type !== "SequenceExpression") ||
+            (seqExpr.expressions.length !== 2)
+          ) {
+            throw tailPath.buildCodeFrameError("Object comprehensions must return" +
+            " sequence expressions of the form `key`, `value`.");
```

Comment:
Been thinking about this a bit, don't think we should say `return` because then people might be tempted to actually write `return`. Also don't think we can say "in tail position" because most people don't know what that means. Do agree that it should change a bit but not sure on final wording.